Index: clang/lib/DirectoryWatcher/windows/DirectoryWatcher-windows.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/clang/lib/DirectoryWatcher/windows/DirectoryWatcher-windows.cpp b/clang/lib/DirectoryWatcher/windows/DirectoryWatcher-windows.cpp
--- a/clang/lib/DirectoryWatcher/windows/DirectoryWatcher-windows.cpp	(revision Staged)
+++ b/clang/lib/DirectoryWatcher/windows/DirectoryWatcher-windows.cpp	(date 1766318679288)
@@ -12,6 +12,7 @@
 #include "llvm/Support/ConvertUTF.h"
 #include "llvm/Support/Path.h"
 #include "llvm/Support/Windows/WindowsSupport.h"
+#include <psapi.h>
 #include <condition_variable>
 #include <mutex>
 #include <queue>
@@ -80,6 +81,98 @@
   void NotifierThreadProc(bool WaitForInitialSync);
 };

+#if _WIN32_WINNT < _WIN32_WINNT_VISTA
+static bool GetFilenameFromHandle(HANDLE hFile, wchar_t *pszFilename)
+{
+  BOOL bRC = FALSE;
+
+  // Get the file size.
+  DWORD dwFileSizeHi = 0;
+  DWORD dwFileSizeLo = GetFileSize(hFile, &dwFileSizeHi);
+
+  // Create a file mapping object.
+  HANDLE hFileMap = CreateFileMappingW(hFile,
+                                      NULL,
+                                      PAGE_READONLY,
+                                      0,
+                                      dwFileSizeLo,
+                                      NULL);
+
+  if (hFileMap)
+  {
+    // Create a file mapping to get the file name.
+    void* pMem = MapViewOfFile(hFileMap, FILE_MAP_READ, 0, 0, 1);
+
+    if (pMem)
+    {
+      if (GetMappedFileNameW(GetCurrentProcess(),
+                            pMem,
+                            pszFilename,
+                            MAX_PATH))
+      {
+        // At this point pszFilename already contains the name.
+        // But the drive name is not written in usual MS DOS fomat, actually,
+        // it is a device name.
+        // So below is an attempt to translate it to drive-letter
+        wchar_t szTemp[512];
+        *szTemp = 0;
+
+        auto uMaxLenDest = wcslen(pszFilename);
+
+        if (GetLogicalDriveStringsW(sizeof (szTemp) - 1, szTemp))
+        {
+          wchar_t szName[MAX_PATH];
+          wchar_t szDrive[3] = L" :";
+          BOOL bFound = FALSE;
+          wchar_t* p = szTemp;
+
+          do
+          {
+            // Copy the drive letter into the template string,
+            // removing the backslash.
+            *szDrive = *p;
+
+            // Look up each device name.
+            if (QueryDosDeviceW(szDrive, szName,
+                               sizeof (szName)))
+            {
+              UINT uNameLen = wcslen(szName);
+
+              //If greater than file name length, it's not a match.
+              if (uNameLen < uMaxLenDest)
+              {
+                bFound = wcsnicmp (pszFilename, szName,
+                                  uNameLen) == 0;
+
+                if (bFound)
+                {
+                  //Reconstruct pszFilename using szTemp as scratch,
+                  // and replace device path with our DOS path.
+                  wchar_t szTempFile[MAX_PATH];
+                  wsprintfW(szTempFile,
+                           L"%s%s",
+                           szDrive,
+                           pszFilename+uNameLen);
+                  wcsncpy(pszFilename, szTempFile, uMaxLenDest);
+                }
+              }
+            }
+            // Go to the next NULL character.
+            while (*p++);
+
+          }
+          while (!bFound && *p); // at the end of the string
+        }
+      }
+      bRC = TRUE;
+      UnmapViewOfFile(pMem);
+    }
+    CloseHandle(hFileMap);
+  }
+  return bRC;
+}
+#endif
+
 DirectoryWatcherWindows::DirectoryWatcherWindows(
     HANDLE DirectoryHandle, bool WaitForInitialSync,
     DirectoryWatcherCallback Receiver)
@@ -87,6 +180,7 @@
   // Pre-compute the real location as we will be handing over the directory
   // handle to the watcher and performing synchronous operations.
   {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
     DWORD Size = GetFinalPathNameByHandleW(DirectoryHandle, NULL, 0, 0);
     std::unique_ptr<WCHAR[]> Buffer{new WCHAR[Size + 1]};
     Size = GetFinalPathNameByHandleW(DirectoryHandle, Buffer.get(), Size, 0);
@@ -97,6 +191,18 @@
       Size -= 4;
     }
     llvm::sys::windows::UTF16ToUTF8(Data, Size, Path);
+#else
+    wchar_t filename[MAX_PATH];
+    wchar_t outPath[2048];
+    wchar_t* Data = outPath;
+    GetFilenameFromHandle(DirectoryHandle, filename);
+    auto Size = wcslen(filename);
+    if (Size >= 4 && ::memcmp(Data, L"\\\\?\\", 8) == 0) {
+      Data += 4;
+      Size -= 4;
+    }
+    llvm::sys::windows::UTF16ToUTF8(Data, Size, Path);
+#endif
   }

   size_t EntrySize = sizeof(FILE_NOTIFY_INFORMATION) + MAX_PATH * sizeof(WCHAR);
Index: llvm/include/llvm/Support/Windows/WindowsSupport.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/llvm/include/llvm/Support/Windows/WindowsSupport.h b/llvm/include/llvm/Support/Windows/WindowsSupport.h
--- a/llvm/include/llvm/Support/Windows/WindowsSupport.h	(revision Staged)
+++ b/llvm/include/llvm/Support/Windows/WindowsSupport.h	(date 1766316274573)
@@ -26,7 +26,7 @@
 #undef _WIN32_IE

 // Require at least Windows 7 API.
-#define _WIN32_WINNT 0x0601
+#define _WIN32_WINNT 0x0501
 #define _WIN32_IE    0x0800 // MinGW at it again. FIXME: verify if still needed.
 #define WIN32_LEAN_AND_MEAN
 #ifndef NOMINMAX
Index: llvm/lib/Support/Windows/Path.inc
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/llvm/lib/Support/Windows/Path.inc b/llvm/lib/Support/Windows/Path.inc
--- a/llvm/lib/Support/Windows/Path.inc	(revision Staged)
+++ b/llvm/lib/Support/Windows/Path.inc	(date 1766318087866)
@@ -27,6 +27,9 @@
 #include "llvm/Support/Windows/WindowsSupport.h"
 #include <shellapi.h>
 #include <shlobj.h>
+#if _WIN32_WINNT < _WIN32_WINNT_VISTA
+#include <psapi.h>
+#endif

 #undef max

@@ -361,9 +364,102 @@
   return is_local_internal(WidePath, result);
 }

+#if _WIN32_WINNT < _WIN32_WINNT_VISTA
+static bool GetFilenameFromHandle(HANDLE hFile, wchar_t *pszFilename)
+{
+  BOOL bRC = FALSE;
+
+  // Get the file size.
+  DWORD dwFileSizeHi = 0;
+  DWORD dwFileSizeLo = GetFileSize(hFile, &dwFileSizeHi);
+
+  // Create a file mapping object.
+  HANDLE hFileMap = CreateFileMappingW(hFile,
+                                      NULL,
+                                      PAGE_READONLY,
+                                      0,
+                                      dwFileSizeLo,
+                                      NULL);
+
+  if (hFileMap)
+  {
+    // Create a file mapping to get the file name.
+    void* pMem = MapViewOfFile(hFileMap, FILE_MAP_READ, 0, 0, 1);
+
+    if (pMem)
+    {
+      if (GetMappedFileNameW(GetCurrentProcess(),
+                            pMem,
+                            pszFilename,
+                            MAX_PATH))
+      {
+        // At this point pszFilename already contains the name.
+        // But the drive name is not written in usual MS DOS fomat, actually,
+        // it is a device name.
+        // So below is an attempt to translate it to drive-letter
+        wchar_t szTemp[512];
+        *szTemp = 0;
+
+        auto uMaxLenDest = wcslen(pszFilename);
+
+        if (GetLogicalDriveStringsW(sizeof (szTemp) - 1, szTemp))
+        {
+          wchar_t szName[MAX_PATH];
+          wchar_t szDrive[3] = L" :";
+          BOOL bFound = FALSE;
+          wchar_t* p = szTemp;
+
+          do
+          {
+            // Copy the drive letter into the template string,
+            // removing the backslash.
+            *szDrive = *p;
+
+            // Look up each device name.
+            if (QueryDosDeviceW(szDrive, szName,
+                               sizeof (szName)))
+            {
+              UINT uNameLen = wcslen(szName);
+
+              //If greater than file name length, it's not a match.
+              if (uNameLen < uMaxLenDest)
+              {
+                bFound = wcsnicmp (pszFilename, szName,
+                                  uNameLen) == 0;
+
+                if (bFound)
+                {
+                  //Reconstruct pszFilename using szTemp as scratch,
+                  // and replace device path with our DOS path.
+                  wchar_t szTempFile[MAX_PATH];
+                  wsprintfW(szTempFile,
+                           L"%s%s",
+                           szDrive,
+                           pszFilename+uNameLen);
+                  wcsncpy(pszFilename, szTempFile, uMaxLenDest);
+                }
+              }
+            }
+            // Go to the next NULL character.
+            while (*p++);
+
+          }
+          while (!bFound && *p); // at the end of the string
+        }
+      }
+      bRC = TRUE;
+      UnmapViewOfFile(pMem);
+    }
+    CloseHandle(hFileMap);
+  }
+  return bRC;
+}
+#endif
+
 static std::error_code realPathFromHandle(HANDLE H,
                                           SmallVectorImpl<wchar_t> &Buffer,
                                           DWORD flags = VOLUME_NAME_DOS) {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
   Buffer.resize_for_overwrite(Buffer.capacity());
   DWORD CountChars = ::GetFinalPathNameByHandleW(
       H, Buffer.begin(), Buffer.capacity(), FILE_NAME_NORMALIZED | flags);
@@ -377,11 +473,20 @@
   Buffer.truncate(CountChars);
   if (CountChars == 0)
     return mapWindowsError(GetLastError());
+#else
+  wchar_t Buff[MAX_PATH];
+  if (!GetFilenameFromHandle(H, Buff)) {
+    return std::error_code();
+  }
+  auto Size = wcslen(Buff);
+  Buffer.resize_for_overwrite(Size);
+#endif
   return std::error_code();
 }

 static std::error_code realPathFromHandle(HANDLE H,
                                           SmallVectorImpl<char> &RealPath) {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
   RealPath.clear();
   SmallVector<wchar_t, MAX_PATH> Buffer;
   if (std::error_code EC = realPathFromHandle(H, Buffer))
@@ -407,6 +512,18 @@
     return EC;

   llvm::sys::path::make_preferred(RealPath);
+#else
+  wchar_t Buff[MAX_PATH];
+  if (!GetFilenameFromHandle(H, Buff)) {
+    return std::error_code();
+  }
+  auto Size = wcslen(Buff);
+  // Convert the result from UTF-16 to UTF-8.
+  if (std::error_code EC = UTF16ToUTF8(Buff, Size, RealPath))
+    return EC;
+
+  llvm::sys::path::make_preferred(RealPath);
+#endif
   return std::error_code();
 }

@@ -421,6 +538,7 @@
 }

 static std::error_code setDeleteDisposition(HANDLE Handle, bool Delete) {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
   // Clear the FILE_DISPOSITION_INFO flag first, before checking if it's a
   // network file. On Windows 7 the function realPathFromHandle() below fails
   // if the FILE_DISPOSITION_INFO flag was already set to 'DeleteFile = true' by
@@ -454,10 +572,14 @@
                                   sizeof(Disposition)))
     return mapWindowsError(::GetLastError());
   return std::error_code();
+#else
+  return errc::not_supported;
+#endif
 }

 static std::error_code rename_internal(HANDLE FromHandle, const Twine &To,
                                        bool ReplaceIfExists) {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
   SmallVector<wchar_t, 0> ToWide;
   if (auto EC = widenPath(To, ToWide))
     return EC;
@@ -481,6 +603,16 @@
   }

   return std::error_code();
+#else
+  wchar_t filename[MAX_PATH];
+  GetFilenameFromHandle(FromHandle, filename);
+  SmallVector<wchar_t, 0> ToWide;
+  if (auto EC = widenPath(To, ToWide))
+    return EC;
+  if (!MoveFileExW(filename, ToWide.begin(), ReplaceIfExists ? MOVEFILE_REPLACE_EXISTING : 0))
+    return mapWindowsError(GetLastError());
+  return std::error_code();
+#endif
 }

 static std::error_code rename_handle(HANDLE FromHandle, const Twine &To) {
@@ -1462,6 +1594,7 @@
 } // end namespace fs

 namespace path {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
 static bool getKnownFolderPath(KNOWNFOLDERID folderId,
                                SmallVectorImpl<char> &result) {
   wchar_t *path = nullptr;
@@ -1474,19 +1607,42 @@
     llvm::sys::path::make_preferred(result);
   return ok;
 }
+#else
+static bool getKnownFolderPath(int folderId, SmallVectorImpl<char> &result) {
+  wchar_t Path[MAX_PATH];
+  if (::SHGetFolderPathW(0, folderId | CSIDL_FLAG_CREATE, 0,
+                         /*SHGFP_TYPE_CURRENT*/ 0, Path) != S_OK)
+    return false;
+  if (UTF16ToUTF8(Path, ::wcslen(Path), result))
+    return false;
+  return true;
+}
+#endif

 bool home_directory(SmallVectorImpl<char> &result) {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
   return getKnownFolderPath(FOLDERID_Profile, result);
+#else
+  return getKnownFolderPath(CSIDL_PROFILE, result);
+#endif
 }

 bool user_config_directory(SmallVectorImpl<char> &result) {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
   // Either local or roaming appdata may be suitable in some cases, depending
   // on the data. Local is more conservative, Roaming may not always be correct.
   return getKnownFolderPath(FOLDERID_LocalAppData, result);
+#else
+  return getKnownFolderPath(CSIDL_LOCAL_APPDATA, result);
+#endif
 }

 bool cache_directory(SmallVectorImpl<char> &result) {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
   return getKnownFolderPath(FOLDERID_LocalAppData, result);
+#else
+return getKnownFolderPath(CSIDL_LOCAL_APPDATA, result);
+#endif
 }

 static bool getTempDirEnvVar(const wchar_t *Var, SmallVectorImpl<char> &Res) {
Index: llvm/lib/Support/Windows/Threading.inc
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/llvm/lib/Support/Windows/Threading.inc b/llvm/lib/Support/Windows/Threading.inc
--- a/llvm/lib/Support/Windows/Threading.inc	(revision Staged)
+++ b/llvm/lib/Support/Windows/Threading.inc	(date 1766317291608)
@@ -137,6 +137,7 @@

 template <typename F>
 static bool IterateProcInfo(LOGICAL_PROCESSOR_RELATIONSHIP Relationship, F Fn) {
+#if _WIN32_WINNT >= _WIN32_WINNT_WIN7
   DWORD Len = 0;
   BOOL R = ::GetLogicalProcessorInformationEx(Relationship, NULL, &Len);
   if (R || GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
@@ -157,9 +158,13 @@
   }
   free(Info);
   return true;
+#else
+  return false;
+#endif
 }

 static std::optional<std::vector<USHORT>> getActiveGroups() {
+#if _WIN32_WINNT >= _WIN32_WINNT_WIN7
   USHORT Count = 0;
   if (::GetProcessGroupAffinity(GetCurrentProcess(), &Count, nullptr))
     return std::nullopt;
@@ -173,6 +178,9 @@
     return std::nullopt;

   return Groups;
+#else
+  return std::nullopt;
+#endif
 }

 static ArrayRef<ProcessorGroup> getProcessorGroups() {
@@ -294,7 +302,7 @@
 // Assign the current thread to a more appropriate CPU socket or CPU group
 void llvm::ThreadPoolStrategy::apply_thread_strategy(
     unsigned ThreadPoolNum) const {
-
+#if _WIN32_WINNT >= _WIN32_WINNT_WIN7
   // After Windows 11 and Windows Server 2022, let the OS do the scheduling,
   // since a process automatically gains access to all processor groups.
   if (llvm::RunningWindows11OrGreater())
@@ -308,9 +316,35 @@
   Affinity.Group = Groups[*Socket].ID;
   Affinity.Mask = Groups[*Socket].Affinity;
   SetThreadGroupAffinity(GetCurrentThread(), &Affinity, nullptr);
+#endif
 }
+
+#if _WIN32_WINNT < _WIN32_WINNT_WIN7
+DWORD GetThreadAffinityMask(HANDLE thread)
+{
+  DWORD Mask = 1;
+  DWORD Old = 0;
+
+  // try every CPU one by one until one works or none are left
+  while(Mask)
+  {
+    Old = SetThreadAffinityMask(thread, Mask);
+    if(Old)
+    {   // this one worked
+      SetThreadAffinityMask(thread, Old);
+      return Old;
+    }
+    if (GetLastError() != ERROR_INVALID_PARAMETER)
+      return 0;
+    Mask <<= 1;
+  }
+
+  return 0;
+}
+#endif

 llvm::BitVector llvm::get_thread_affinity_mask() {
+#if _WIN32_WINNT >= _WIN32_WINNT_WIN7
   GROUP_AFFINITY Affinity{};
   GetThreadGroupAffinity(GetCurrentThread(), &Affinity);

@@ -330,6 +364,16 @@
       V.set(StartOffset + I);
   }
   return V;
+#else
+  DWORD_PTR Affinity = GetThreadAffinityMask(GetCurrentThread());
+  llvm::BitVector V;
+  V.resize(sizeof(DWORD));
+  for (unsigned I = 0; I < sizeof(DWORD); ++I) {
+    if ((Affinity >> I) & 1)
+      V.set(I);
+  }
+  return V;
+#endif
 }

 unsigned llvm::get_cpus() { return getProcessorGroups().size(); }
Index: llvm/lib/Support/raw_socket_stream.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/llvm/lib/Support/raw_socket_stream.cpp b/llvm/lib/Support/raw_socket_stream.cpp
--- a/llvm/lib/Support/raw_socket_stream.cpp	(revision Staged)
+++ b/llvm/lib/Support/raw_socket_stream.cpp	(date 1766316906912)
@@ -191,6 +191,7 @@
 manageTimeout(const std::chrono::milliseconds &Timeout,
               const std::function<int()> &getActiveFD,
               const std::optional<int> &CancelFD = std::nullopt) {
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
   struct pollfd FD[2];
   FD[0].events = POLLIN;
 #ifdef _WIN32
@@ -248,6 +249,7 @@
     return std::make_error_code(std::errc::timed_out);
   if (FD[0].revents & POLLNVAL)
     return std::make_error_code(std::errc::bad_file_descriptor);
+#endif
   return std::error_code();
 }

Index: llvm/tools/llvm-ar/llvm-ar.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/llvm/tools/llvm-ar/llvm-ar.cpp b/llvm/tools/llvm-ar/llvm-ar.cpp
--- a/llvm/tools/llvm-ar/llvm-ar.cpp	(revision Staged)
+++ b/llvm/tools/llvm-ar/llvm-ar.cpp	(date 1766318608461)
@@ -562,12 +562,17 @@
 // CompareStringOrdinal compares two Unicode strings for
 // binary equivalence and allows for case insensitivity.
 #ifdef _WIN32
+
   SmallVector<wchar_t, 128> WPath1, WPath2;
   failIfError(sys::windows::UTF8ToUTF16(normalizePath(Path1), WPath1));
   failIfError(sys::windows::UTF8ToUTF16(normalizePath(Path2), WPath2));
-
+#if _WIN32_WINNT >= _WIN32_WINNT_VISTA
   return CompareStringOrdinal(WPath1.data(), WPath1.size(), WPath2.data(),
                               WPath2.size(), true) == CSTR_EQUAL;
+#else
+  return CompareStringW(LOCALE_INVARIANT,0, WPath1.data(), WPath1.size(), WPath2.data(),
+                              WPath2.size()) == CSTR_EQUAL;
+#endif
 #else
   return normalizePath(Path1) == normalizePath(Path2);
 #endif
