diff --git a/src/binutils.mk b/src/binutils.mk
index e26baeaa..88ab62e4 100644
--- a/src/binutils.mk
+++ b/src/binutils.mk
@@ -7,7 +7,7 @@ $(PKG)_VERSION  := 2.38
 $(PKG)_CHECKSUM := 070ec71cf077a6a58e0b959f05a09a35015378c2d8a51e90f3aeabfe30590ef8
 $(PKG)_SUBDIR   := binutils-$($(PKG)_VERSION)
 $(PKG)_FILE     := binutils-$($(PKG)_VERSION).tar.bz2
-$(PKG)_URL      := https://ftp.gnu.org/gnu/binutils/$($(PKG)_FILE)
+$(PKG)_URL      := https://ftpmirror.gnu.org/binutils/$($(PKG)_FILE)
 $(PKG)_URL_2    := https://ftpmirror.gnu.org/binutils/$($(PKG)_FILE)
 $(PKG)_DEPS     :=

diff --git a/src/gcc.mk b/src/gcc.mk
index 6e69ff6e..f93afc4f 100644
--- a/src/gcc.mk
+++ b/src/gcc.mk
@@ -9,7 +9,7 @@ $(PKG)_RELEASE  := $($(PKG)_VERSION)
 $(PKG)_CHECKSUM := a6e21868ead545cf87f0c01f84276e4b5281d672098591c1c896241f09363478
 $(PKG)_SUBDIR   := gcc-$($(PKG)_VERSION)
 $(PKG)_FILE     := gcc-$($(PKG)_VERSION).tar.xz
-$(PKG)_URL      := https://ftp.gnu.org/gnu/gcc/gcc-$($(PKG)_VERSION)/$($(PKG)_FILE)
+$(PKG)_URL      := https://www.mirrorservice.org/sites/sourceware.org/pub/gcc/releases/gcc-$($(PKG)_VERSION)/$($(PKG)_FILE)
 $(PKG)_URL_2    := https://www.mirrorservice.org/sites/sourceware.org/pub/gcc/releases/gcc-$($(PKG)_VERSION)/$($(PKG)_FILE)
 $(PKG)_DEPS     := binutils mingw-w64 $(addprefix $(BUILD)~,gmp isl mpc mpfr)

